motoman_driver/src/io_relay.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ * Copyright (c) 2013, Southwest Research Institute
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright
+ *  notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *  notice, this list of conditions and the following disclaimer in the
+ *  documentation and/or other materials provided with the distribution.
+ *  * Neither the name of the Southwest Research Institute, nor the names
+ *  of its contributors may be used to endorse or promote products derived
+ *  from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "motoman_driver/io_relay.h"
+#include <string>
+#include <ros/ros.h>
+
+using industrial::shared_types::shared_int;
+
+namespace motoman
+{
+namespace io_relay
+{
+
+// TODO: is this needed?
+#define ROS_ERROR_RETURN(rtn,...) do {ROS_ERROR(__VA_ARGS__); return(rtn);} while(0)
+
+bool MotomanIORelay::init(std::string default_ip, int default_port)
+{
+  ROS_INFO("MotomanIORelay: init");
+
+  std::string ip;
+  int port;
+
+  // override IP/port with ROS params, if available
+  ros::param::param<std::string>("robot_ip_address", ip, default_ip);
+  ros::param::param<int>("~port", port, default_port);
+
+  // check for valid parameter values
+  if (ip.empty())
+  {
+    ROS_ERROR("No valid robot IP address found.  Please set ROS 'robot_ip_address' param");
+    return false;
+  }
+
+  char* ip_addr = strdup(ip.c_str());  // connection.init() requires "char*", not "const char*"
+  ROS_INFO("I/O relay connecting to IP address: '%s:%d'", ip_addr, port);
+  default_tcp_connection_.init(ip_addr, port);
+  free(ip_addr);
+  default_tcp_connection_.makeConnect();
+
+  io_ctrl_.init(&default_tcp_connection_);
```

Comment:
Can `init(..)` fail? If so: should we check its return value?

Reply:
It looks like the `MotomanIOCtrl::init` function is currently set to always return true. Should I change the function type to void? Or should I leave it how it is and add in a check here for the bool (in case we decide to add error checking to the `init` function later)?

https://github.com/steviedale/motoman/blob/8729f66b5fccb1784e7c7c62706ffb7a978f0f32/motoman_driver/src/io_ctrl.cpp#L62-L66

Reply:
As the method returns a `bool`, that's what we should check. Fact it cannot be anything but `true` right now doesn't change that.

Reply:
Addressed in abed8fd

Sounds good. I also added a similar check to the `TcpClient.init` call.